src/main/java/com/tabnine/Initializer.kt

```diff
@@ -0,0 +1,88 @@
+package com.tabnine
+
+import com.intellij.ide.plugins.PluginInstaller
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.application.PreloadingActivity
+import com.intellij.openapi.components.ServiceManager
+import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.progress.ProgressIndicator
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.startup.StartupActivity
+import com.tabnine.capabilities.CapabilitiesService
+import com.tabnine.config.Config
+import com.tabnine.dialogs.Dialogs.showRestartDialog
+import com.tabnine.dialogs.TabnineEnterpriseUrlDialogWrapper
+import com.tabnine.general.DependencyContainer
+import com.tabnine.general.StaticConfig
+import com.tabnine.general.Utils
+import com.tabnine.lifecycle.BinaryNotificationsLifecycle
+import com.tabnine.lifecycle.BinaryPromotionStatusBarLifecycle
+import com.tabnine.lifecycle.BinaryStateService
+import com.tabnine.lifecycle.TabnineUpdater
+import com.tabnine.logging.initTabnineLogger
+import com.tabnine.notifications.ConnectionLostNotificationHandler
+import com.tabnine.userSettings.AppSettingsState.Companion.instance
+import java.util.concurrent.atomic.AtomicBoolean
+
+class Initializer : PreloadingActivity(), StartupActivity {
+    private var binaryNotificationsLifecycle: BinaryNotificationsLifecycle? = null
+    private var binaryPromotionStatusBarLifecycle: BinaryPromotionStatusBarLifecycle? = null
+    override fun preload(indicator: ProgressIndicator) {
+        initialize()
+    }
+
+    override fun runActivity(project: Project) {
+        initialize()
+    }
+
+    private fun initialize() {
```

Comment:
1. i think the `dialogShown` boolean is redundant, I don't see any reason why not to use the `initialized` atomic boolean for this - i.e. if not initizlied && on prem => show dialog
2. the flow of this function would be much clearer IMO like so:
```kotlin
if (!shouldInitialize) return
// ... rest of the code
if (Config.IS_ON_PREM) {
  requireUserCloud2Url();
} else {
  initializeNonOnPremListeners();
}
```
where `requireUserCloud2Url` will contain all the `if` and `else` clauses of the `dialogShwon` related code (i.e. show the dialog if needed and set the custom repo URL)

and `initializeNonOnPremListeners` will contain all the code under the `!Config.IS_ON_PREM` condition in the current code